Support hardware checksum offloading, only available in zeek 3.2+

timwoj opened the pull request:
I've had this patch sitting in a fork for a while and @sethhall reminded me that I never actually opened a PR to upstream it.

This PR adds support for handling hardware checksumming from af_packet and passing that information to zeek to allow it to skip the internal checksums.

J-Gras commented:
Thanks a lot! I guess this is already tested, right? Do you think it would make sense to make this configurable? I also thought about combining this with the VLAN support in 9178e5a313d789cf38e51c1f2dafda881b942a45.

timwoj commented:
I tested it briefly when I added the `l3_checksummed` variables in Zeek, but not much beyond that. I agree that it should probably have an option to make it configurable.

timwoj commented:
I have an Intel NUC here running Ubuntu 22.04. I enabled hardware checksumming via:

```
ethtool -K eno1 rx on
ethtool -K eno1 tx on
```

I also added a little bit of debugging logging the change here just to output that we're getting the flags in the data. The machine doesn't really get a lot of network traffic, but I tickled it a bit by doing some http requests, dns queries, ssh connections, etc, and it looks like it's working as expected. I'm getting proper log entries for everything I tried, and I'm getting output from the logging I added. Out of 1725ish packets, roughly 95% had a good checksum flag. The other 5% were either missing the flag, the flag was set to bad, or the flag was set to not ready.

J-Gras commented:
Just for the record, we have two open questions regarding this:

1. The semantics of `TP_STATUS_CSUM_VALID` and `TP_STATUS_CSUMNOTREADY` might be fuzzy (see [LKML discussion](https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg2182237.html)). Current assumption: `TP_STATUS_CSUM_VALID` means that IP checksum is valid if required. TCP checksums can still be bad (maybe only in tunneling scenarios). `TP_STATUS_CSUMNOTREADY` signals that checksums are not necessary as the packet never left the machine.
2. Zeek's naming of `l2_checksummed` and `l3_checksummed` seems confusing as it is used for IP and TCP respectively. However, in other places in Zeek L3 refers to the IP layer.

timwoj commented:
> The semantics of TP_STATUS_CSUM_VALID and TP_STATUS_CSUMNOTREADY might be fuzzy (see [LKML discussion](https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg2182237.html)).

From how I understand that LKML posting, `TP_STATUS_CSUM_VALID` basically means that the TCP checksum is correct. That would mean we can, at the very least, set `l4_checksummed` to true and skip that step in Zeek. If you're running a new enough kernel, and the new `TP_STATUS_CSUM_UNNECESSARY` is available, you can also set `l2_checksummed` and `l3_checksummed` to true since those should have been validated as well.

> Zeek's naming of l2_checksummed and l3_checksummed seems confusing as it is used for IP and TCP respectively. However, in other places in Zeek L3 refers to the IP layer.

This part has been fixed in Zeek's master branch now. I'll see about backporting it for the next 5.0.x release (which should be soonish, probably).

> - receiver uses nfp (netronome) driver: TP_STATUS_CSUM_VALID set for every packet, including the bad TCP ones

This part of that LKML thread seems to indicate that there might be NIC drivers out there that don't handle the checksum flags correctly, unless I'm misreading that part (https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg2182927.html). Unless there's some way to detect that, this whole thing might be dead in the water.

And then you get later in the thread (https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg2182944.html) and it seems like `TP_STATUS_CSUM_VALID` may indicate that all of the transport layer checksums are valid, even ones inside of tunnels. 

I will definitely agree that the whole thing is horribly documented.

timwoj commented:
I added that script-level variable that we talked about. The plugin already had a bunch of other variables that followed that same model. I also rebased it while I was at it since there were some other changes on the upstream version recently.

awelzel commented:
Food for thought around naming and type of the script-layer var: The suricata config uses `checksum-validation` and has values  `kernel`, `off`, `on`, `auto`. For this plugin, only `kernel` (trust TP_STATUS_CSUM_VALID) and `on` (let Zeek compute them).

https://github.com/OISF/suricata/blob/a9a17c81855e219ae27f94b803a29d74b6facb37/suricata.yaml.in#L641-L651

Using an enum instead of the bool would allow future extensibility and not make the C++ too painful. Having `hw` in the name I'm also a little unsure about - the checksums may/could have been validated in software as well before being handed to user-space. I like Suricata's choice of "kernel" here). So, I could see `redef AF_Packet::checksum_validation = KERNEL;` or so.

awelzel commented on CMakeLists.txt:
```diff
@@ -13,6 +13,13 @@ zeek_plugin_bif(src/af_packet.bif)
 zeek_plugin_dist_files(zeekctl/af_packet.py README COPYING VERSION)
 zeek_plugin_end()
 
+if ( AF_PACKET_HARDWARE_OFFLOAD )
```
Agree with your comment, probably not useful if there's a variable and no need for the configure flag.

The name's also too generic not mentioning checksum, but anyway, would suggest to just remove.

timwoj commented on src/AF_Packet.cc:
```diff
@@ -250,6 +250,15 @@ bool AF_PacketSource::ExtractNextPacket(zeek::Packet* pkt)
 			pkt->vlan = packet->hv1.tp_vlan_tci;
 
+#if defined(AF_PACKET_HARDWARE_OFFLOAD) && ZEEK_VERSION_NUMBER >= 50100
+		if ( zeek::BifConst::AF_Packet::enable_hw_timestamping &&
```
In my defense, I wrote this sitting in an airport being annoyed that they kept delaying my filght multiple times and moving the gate around the airport.

J-Gras commented:
> Using an enum instead of the bool would allow future extensibility and not make the C++ too painful.

I like that approach. In particular, the emphazise on `kernel` makes sense and might encourage users to scrutinize the semantics. In addition to `kernel` and `on`, `off` could be implemented just by setting `l{2-4}_checksummed` to true.

timwoj commented:
> I like that approach. In particular, the emphazise on `kernel` makes sense and might encourage users to scrutinize the semantics. In addition to `kernel` and `on`, `off` could be implemented just by setting `l{2-4}_checksummed` to true.

Yeah, I like this idea too. Commit incoming.
